param.go

```diff
@@ -207,7 +217,44 @@ func (ps paramSingle) getValue(c containerStore) (reflect.Value, bool) {
 	return _noValue, false
 }
 
-func (ps paramSingle) Build(c containerStore) (reflect.Value, error) {
+// builds the parameter using decorators, if any. If there are no decorators associated
+// with this parameter, this returns _noValue.
+func (ps paramSingle) buildWithDecorators(c containerStore) (v reflect.Value, err error, found bool) {
+	decorators := c.getValueDecorators(ps.Name, ps.Type)
+	if len(decorators) != 0 {
+		found = true
+		for _, d := range decorators {
+			err := d.Call(c)
+			if err == nil {
+				continue
+			}
+			if _, ok := err.(errMissingDependencies); ok && ps.Optional {
+				v, err = reflect.Zero(ps.Type), nil
+			}
+			return _noValue, errParamSingleFailed{
+				CtorID: 1,
+				Key:    key{t: ps.Type, name: ps.Name},
+				Reason: err,
+			}, true
+		}
+		v, _ = c.getDecoratedValue(ps.Name, ps.Type)
+	}
+	return v, err, found
```

Comment:
Can reduce nesting here by inverting the if condition
```suggestion
	if len(decorators) == 0 {
		return _noValue, nil, false
	}

	for _, d := range decorators {
		err := d.Call(c)
		if err == nil {
			continue
		}
		if _, ok := err.(errMissingDependencies); ok && ps.Optional {
			v, err = reflect.Zero(ps.Type), nil
		}
		return _noValue, errParamSingleFailed{
			CtorID: 1,
			Key:    key{t: ps.Type, name: ps.Name},
			Reason: err,
		}, true
	}
	v, _ = c.getDecoratedValue(ps.Name, ps.Type)
	return v, err, true
```